test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"time"
+
+	"golang.org/x/oauth2/google"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	. "github.com/onsi/ginkgo"
+	"io"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	instrumentation "k8s.io/kubernetes/test/e2e/instrumentation/common"
+	"reflect"
+)
+
+const (
+	// Time to wait after a pod creation for it's metadata to be exported
+	metadataWaitTime = 120 * time.Second
+
+	// Scope for Stackdriver Metadata API
+	MonitoringScope = "https://www.googleapis.com/auth/monitoring"
+)
+
+var _ = instrumentation.SIGDescribe("Stackdriver Monitoring", func() {
+	BeforeEach(func() {
+		framework.SkipUnlessProviderIs("gce", "gke")
+	})
+
+	f := framework.NewDefaultFramework("stackdriver-monitoring")
+	var kubeClient clientset.Interface
+
+	It("should run Stackdriver Metadata Agent [Feature:StackdriverMetadataAgent]", func() {
+		kubeClient = f.ClientSet
+		testAgent(f, kubeClient)
+	})
+})
+
+func testAgent(f *framework.Framework, kubeClient clientset.Interface) {
+	projectId := framework.TestContext.CloudConfig.ProjectID
+	resourceType := "k8s_container"
+	uniqueContainerName := fmt.Sprintf("test-container-%v", time.Now().Unix())
+	endpoint := fmt.Sprintf(
+		"https://stackdriver.googleapis.com/v1beta2/projects/%v/resourceMetadata?filter=resource.type%%3D%v+AND+resource.label.container_name%%3D%v",
+		projectId,
+		resourceType,
+		uniqueContainerName)
+
+	oauthClient, err := google.DefaultClient(context.Background(), MonitoringScope)
+	if err != nil {
+		framework.Failf("Failed to create oauth client: %s", err)
+	}
+
+	// Create test pod with unique name.
+	framework.CreateExecPodOrFail(kubeClient, f.Namespace.Name, uniqueContainerName, func(pod *v1.Pod) {
+		pod.Spec.Containers[0].Name = uniqueContainerName
+	})
+	defer kubeClient.CoreV1().Pods(f.Namespace.Name).Delete(uniqueContainerName, &metav1.DeleteOptions{})
+
+	// Wait a short amount of time for Metadata Agent to be created and metadata to be exported
+	time.Sleep(metadataWaitTime)
+
+	resp, err := oauthClient.Get(endpoint)
+	if err != nil {
+		framework.Failf("Failed to call Stackdriver Metadata API %s", err)
+	}
+	if resp.StatusCode != 200 {
+		framework.Failf("Stackdriver Metadata API returned error status: %s", resp.Status)
+	}
+	metadataAPIResponse := new(bytes.Buffer)
```

Comment:
When I saw this, `ioutil.ReadAll` came to mind. Would using that make this code simpler if we're only using the buffer to extract `.Bytes()`?

https://golang.org/pkg/io/ioutil/#ReadAll

Reply:
Done